rfcs/012-session-persistence/012-session-persistence.md

```diff
@@ -0,0 +1,303 @@
+
+- … many more 😅
+
+### Summary
+
+The current implementation of *saving* work when a hosted user session shuts down leads to confusion in the best-, and
+data loss in the worst-case scenarios. There is a need for a **workspace** in a project that users can come to where their
+data and progress are reliably stored.
+
+## Key Assumptions
+
+- We assume that there is one session per project per user.
+
+## Proposed solution
+
+From the user's perspective, connecting to a session, quitting, and reconnecting again is like coming and going from a
+**workspace** - they don't think about it in terms of the underlying git repository state, and we should not expect them
+to. If the state of their current workspace (session volume) is different from what they asked for, they are presented
+with options to make an informed and deliberate decision about how to proceed. A good template for the behavior we
+should be going for is the UX of GitHub Codespaces.
+
+### TL;DR of the Technical Proposal
+
+1. Rely on Permanent Volumes for saving session state.
+2. On session shutdown, transition the session pod to a minimal state with PV and helper containers to communicate
+   about, and change the state of the repo on-disk.
+3. On session launch, check if an existing PV for the project/user exists - if yes, compare the requested state with the
+   state on disk and inform the user about a potential mismatch.
+
+## Technical Details
+
+We want to eliminate *autosave branches* as a way to store session state. This choice has some consequences:
+
+- If sessions are to be resilient against failures of any kind, they *must* be using Permanent Volumes for session
+  storage.
+
+- To simplify PV management, we assume there is *one session per project per user*; we believe that if the user wants to
+  compare different states of their project, they will simply use `git` inside the session for this. Note that this
+  assumption may need to be revisited at a later point, but we think it's a good tradeoff given the rising complexity of
+  managing sessions.
+
+The basic idea then is to:
+
+1. Rely on the PV mechanism to preserve user data; this has the benefit of being resilient to various timeouts and
+   synchronization issues of autosave branches as well as eliminating unnecessary objects polluting LFS.
+2. Inform clients about the state of the repository *inside the PV* and to manipulate it to match what the user wants. 
+3. Present the user with simple choices when the requested session does not match the state inside the PV; allow users
+   to take informed decisions when there is no obvious way forward
+
+### Additional Technical Details
+
+The main difference that we would introduce in Amalthea/Sessions is the following:
+
+1. Sessions exist in one of 3 forms:
+   - **Running**: This is the same as the current sessions
+       - Containers: `oauth2-proxy`, `git-proxy`, `sidecar`, `session`
+       - No changes from currently running sessions
+   - **Hibernate**:
```

Comment:
Alternative wording: "Sleeping".

Reply:
I think hibernating fits very well in this context and most people liked it. `Sleeping` is also fine but let's stick to the word that we agreed upon.